src/chainstate/burn/db/sortdb.rs

```diff
@@ -2254,13 +2397,85 @@ impl <'a> SortitionHandleTx <'a> {
+                    for (ix, address) in reward_set.iter().enumerate() {
+                        keys.push(db_keys::pox_reward_set_entry(ix as u16));
+                        values.push(address.to_string());
+                    }
+                } else {
+                    keys.push(db_keys::pox_reward_set_size().to_string());
+                    values.push(db_keys::reward_set_size_to_string(0));
+                }
+
+                // in all cases, write the new PoX bit vector
                 keys.push(db_keys::pox_identifier().to_string());
                 values.push(pox_id.to_string());
+            } else {
+                // if this snapshot consumed some reward set entries AND
+                //  this isn't the start of a new reward cycle,
+                //   update the reward set
+                if let Some(reward_info) = recipient_info {
+                    let mut current_len = self.get_reward_set_size()?;
+                    let mut recipient_indexes: Vec<_> = reward_info.recipients.iter().map(|(_, x)| *x).collect();
+                    let mut remapped_entries = HashMap::new();
+                    // sort in decrementing order
+                    recipient_indexes.sort_unstable_by(|a, b| b.cmp(a));
+                    for index in recipient_indexes.into_iter() {
+                        // sanity check
+                        if index >= current_len {
+                            unreachable!("Supplied index should never be greater than recipient set size");
+                        } else if index + 1 == current_len {
+                            // selected index is the last element: no need to swap, just decrement len
+                            current_len -= 1;
+                        } else {
+                            let replacement = current_len - 1; // if current_len were 0, we would already have panicked.
+                            let replace_with =
+                                if let Some((_prior_ix, replace_with)) = remapped_entries.remove_entry(&replacement) {
+                                    // the entry to swap in was itself swapped, so let's use the new value instead
+                                    replace_with
+                                } else {
+                                    self.get_reward_set_entry(replacement)?
+                                };
+
+                            // swap and decrement to remove from set
+                            remapped_entries.insert(index, replace_with);
+                            current_len -= 1;
+                        }
+                    }
+                    // store the changes in the new trie
+                    keys.push(db_keys::pox_reward_set_size().to_string());
+                    values.push(db_keys::reward_set_size_to_string(current_len as usize));
+                    for (index, new_address) in remapped_entries.into_iter() {
```

Comment:
The ordering of the pairs in `remapped_entries` is not guaranteed to be stable across nodes, since the hash map implementation may use different hash functions.  Is this okay?

Reply:
Yeah, that's okay here -- these keys are all unique, and any differences in the sortition MARFs due to ordering won't matter because nodes don't use the sortition MARF root hash for anything.

Reply:
The ordering of the reward set when it is first inserted, though, _definitely_ matters, and that's something we should probably perform a sort over before insertion.

Reply:
In the interest of moving this PR along, I've made an issue to track this: https://github.com/blockstack/stacks-blockchain/issues/1860